[iOS] Use unique temporary directory for each WebKit process type

pvollan opened the pull request:
#### 85dbd47a744cae233f19054afa9e7fe89870a119
<pre>
[iOS] Use unique temporary directory for each WebKit process type
<a href="https://bugs.webkit.org/show_bug.cgi?id=291847">https://bugs.webkit.org/show_bug.cgi?id=291847</a>
<a href="https://rdar.apple.com/149690496">rdar://149690496</a>

Reviewed by Sihui Liu.

Use a unique temporary folder for each WebKit process type in the parent process&apos; data container.
For the Networking process, we create a read extension to the root of the temporary directory in
the app&apos;s container, since some apps rely on the Networking process having read access there when
loading data with subresources located in the temporary directory.

This patch was previously landed in 293291@main, but was reverted because some apps relied on the
Networking process to have read access to the root of the temporary directory.

* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::addWebsiteDataStore):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
(WebKit::setUserDirSuffix):
(WebKit::XPCServiceEventHandler):
* Source/WebKit/Shared/WebKitServiceNames.h: Added.
* Source/WebKit/Shared/WebProcessDataStoreParameters.h:
* Source/WebKit/Shared/WebProcessDataStoreParameters.serialization.in:
* Source/WebKit/Shared/WebsiteDataStoreParameters.h:
* Source/WebKit/Shared/WebsiteDataStoreParameters.serialization.in:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::GPUProcessProxy):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::webProcessDataStoreParameters):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::parameters):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::platformSetWebsiteDataStoreParameters):

Canonical link: <a href="https://commits.webkit.org/294059@main">https://commits.webkit.org/294059@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e97326382f06ac65c873d7ba476064869e34dbb5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100521 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20173 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10472 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105658 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51109 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20481 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28647 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76549 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33590 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103528 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15708 "Found 1 new test failure: fast/forms/ios/file-upload-panel-dismiss-when-view-removed-from-window.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90805 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56906 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15525 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50485 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85438 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8887 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108012 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27639 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20293 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85501 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28002 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87005 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85039 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29731 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7459 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21610 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16380 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27574 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32824 "Found 32 new failures in platform/audio/PlatformMediaSessionInterface.h, platform/audio/PlatformMediaSession.cpp") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27385 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30703 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28943 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -886,9 +887,18 @@ WebProcessDataStoreParameters WebProcessPool::webProcessDataStoreParameters(WebP
+    SandboxExtension::Handle containerTemporaryDirectoryExtensionHandle;
+    SandboxExtension::Handle containerTemporaryDirectoryExtensionHandleForWebContent;
+    if (auto directory = websiteDataStore.resolvedContainerTemporaryDirectory(); !directory.isEmpty()) {
+        if (WTF::IOSApplication::isGmail()) {
```
Could you add radar link?

pvollan replied:
After Slack discussion, we removed this quirk.

Thanks for reviewing!

dewei-zhu commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -603,8 +603,8 @@ static void setVideoDecoderBehaviors(OptionSet<VideoDecoderBehavior> videoDecode
-    if (auto& handle = parameters.containerTemporaryDirectoryExtensionHandle)
-        grantAccessToContainerTempDirectory(*handle);
+    SandboxExtension::consumePermanently(parameters.containerTemporaryDirectoryExtensionHandle);
+    grantAccessToContainerTempDirectory(parameters.containerTemporaryDirectoryExtensionHandleForWebContent);
```
`grantAccessToContainerTempDirectory(parameters.containerTemporaryDirectoryExtensionHandleForWebContent);` does not check if `parameters.containerTemporaryDirectoryExtensionHandle` is evaluated true, we probably need to do it in  `grantAccessToContainerTempDirectory` or keep `if (auto& handle = parameters.containerTemporaryDirectoryExtensionHandle)` check.

dewei-zhu replied:
NVM, it looks like 
`std::optional<SandboxExtension::Handle> containerTemporaryDirectoryExtensionHandle` is changed to 
```
SandboxExtension::Handle containerTemporaryDirectoryExtensionHandle;
SandboxExtension::Handle containerTemporaryDirectoryExtensionHandleForWebContent;
```

pvollan replied:
Ah, yes, good point, we changed this to `SandboxExtension::Handle`.

Thanks for reviewing!

pvollan committed:
[iOS] Use unique temporary directory for each WebKit process type
https://bugs.webkit.org/show_bug.cgi?id=291847
rdar://149690496

Reviewed by Sihui Liu.

Use a unique temporary folder for each WebKit process type in the parent process' data container.
For the Networking process, we create a read extension to the root of the temporary directory in
the app's container, since some apps rely on the Networking process having read access there when
loading data with subresources located in the temporary directory.

This patch was previously landed in 293291@main, but was reverted because some apps relied on the
Networking process to have read access to the root of the temporary directory.

* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::addWebsiteDataStore):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
(WebKit::setUserDirSuffix):
(WebKit::XPCServiceEventHandler):
* Source/WebKit/Shared/WebKitServiceNames.h: Added.
* Source/WebKit/Shared/WebProcessDataStoreParameters.h:
* Source/WebKit/Shared/WebProcessDataStoreParameters.serialization.in:
* Source/WebKit/Shared/WebsiteDataStoreParameters.h:
* Source/WebKit/Shared/WebsiteDataStoreParameters.serialization.in:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::GPUProcessProxy):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::webProcessDataStoreParameters):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::parameters):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::platformSetWebsiteDataStoreParameters):

Canonical link: https://commits.webkit.org/294059@main
